### PR TITLE
Double precision

### PR DESCRIPTION
**Pull request by delandmeterp**
This PR solves a numerical issue that we observe with Parcels.

In coastal flows, the particles tend to follow the coastlines, and can be very close to the coast. Impermeable boundary condition of the input data combined with a small time step (~15 minutes) should prevent the particles of beaching, but rounding errors result in beaching particles. By converting lon, lat, depth to double precision, we reduce this un-controlled process by ~95%.

So it can have a large effect on the particle general dynamics.

How could we solve this problem differently?

* Using unbeaching kernels (they have a variable effect depending on mesh resolution)
* Hacking the interpolation scheme (if UV is zero, and I am just on the edge of the beach, I put back the particle on the other side of the edge)
* By simply putting double precision as a flag of the particle set.

Any suggestion on the solution to adopt?

### CODE REVIEW

**Comment by delandmeterp** [comment]
About commit [694c8df]

It seems that we had to increase a lot the tolerance to pass the test. It is not really the case.
Indeed, we were comparing with a very small tolerance numbers stored as float, which resulted before in a difference of 0.

See attached code for a simple example: `a` is considered equal to `b` because the tolerance is smaller than float accuracy.
```
#include <stdio.h>
#include <math.h>

int main(){
  float a = 4.5;
  a += 1e-8;
  float b = 4.5;
  if (fabs(a-b) < 1e-12)
    printf("a == b\n");
  else
    printf("a !=  b\n");

  return 0;
}
```

**Comment by willirath** [comment]
There's considerable efforts in the modelling community to explore reduced precision, because this will yield much higher performance on, e.g., GPUs.  Just defaulting to double is contrary to this tendency.

Would it make sense to consciously handle precision and formulate tolerances in terms of machine precision?

**Comment by delandmeterp** [comment]
Indeed and we try to avoid using double precision where it's not necessary. Then why this change?

It solves a numerical issue. In coastal flows, the particles tend to follow the coastlines, and can be very close to the coast. Impermeable boundary condition of the input data combined with a small time step (~15 minutes) should prevent the particles of beaching, but rounding errors result in beaching particles. By converting lon, lat, depth to double precision, we reduce this un-controlled process by ~95%.

So it can have a large effect on the particle general dynamics.

How could we solve this problem differently?
 * Using unbeaching kernels (they have a variable effect depending on mesh resolution)
 * Hacking the interpolation scheme (if UV is zero, and I am just on the edge of the beach, I put back the particle on the other side of the edge)
 * By simply putting double precision as a flag of the particle set.

Any suggestion on the solution to adopt?

File: parcels/particleset.py [hunk 1]
```diff
@@ -29,13 +29,13 @@ class ParticleSet(object):
     :param depth: Optional list of initial depth values for particles. Default is 0m
     :param time: Optional list of initial time values for particles. Default is fieldset.U.grid.time[0]
     :param repeatdt: Optional interval (in seconds) on which to repeat the release of the ParticleSet
-    :param coordinates_var_precision: Floating precision for lon, lat, depth particle coordinates.
+    :param lonlatdepth_dtype: Floating precision for lon, lat, depth particle coordinates.
            It is either 'single' or 'double'. Default is 'single' if fieldset.U.interp_method is 'linear'
```
Comment (erikvansebille) [inline parcels/particleset.py]: Shouldn't we then also rename the options to `np.float32` and `np.float64`?

File: parcels/kernels/advection.py [hunk 2]
```diff
@@ -57,8 +57,8 @@ def AdvectionRK45(particle, fieldset, time):
 
     Times-step dt is halved if error is larger than tolerance, and doubled
     if error is smaller than 1/10th of tolerance, with tolerance set to
-    1e-9 * dt by default."""
-    tol = [1e-9]
+    1e-5 * dt by default."""
+    tol = [1e-5]
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply parcels/kernels/advection.py]: I would keep this tolerance, which is actually usable with double precision.
The other one either fails (in double) or is way below what is caught by single floating precision.

File: parcels/particleset.py [hunk 3]
```diff
@@ -82,6 +86,21 @@ def convert_to_list(var):
...
+                for f in fieldset.U:
+                    if f.interp_method == 'cgrid_velocity':
+                        self.coordinates_var_precision = 'double'
+                        break
+            else:
+                self.coordinates_var_precision = 'double' if fieldset.U.interp_method == 'cgrid_velocity' else 'single'
+        else:
+            self.coordinates_var_precision = coordinates_var_precision
+        assert self.coordinates_var_precision in ['single', 'double'], \
+            'particle coordinate variable precision is either set at single or double'
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply parcels/particleset.py]: done

File: parcels/particleset.py [hunk 4]
```diff
@@ -143,8 +162,10 @@ def from_line(cls, fieldset, pclass, start, finish, size, depth=None, time=None,
         :param time: Optional start time value for particles. Default is fieldset.U.time[0]
         :param repeatdt: Optional interval (in seconds) on which to repeat the release of the ParticleSet
         """
-        lon = np.linspace(start[0], finish[0], size, dtype=np.float32)
-        lat = np.linspace(start[1], finish[1], size, dtype=np.float32)
+
+        lonlat_type = np.float64 if fieldset.U.interp_method == 'cgrid_velocity' else np.float32
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply parcels/particleset.py]: done

File: tests/test_fieldset.py [hunk 5]
```diff
@@ -436,7 +436,7 @@ def generate_dataset(xdim, ydim, zdim=1, tdim=1):
     pset = ParticleSet(fieldset, JITParticle, 0, 0)
 
     pset.execute(AdvectionRK4, dt=1)
-    assert pset[0].lon == 4.5 and pset[0].lat == 10
+    assert np.allclose(pset[0].lon, 4.5) and np.allclose(pset[0].lat, 10)
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply tests/test_fieldset.py]: I would say it's never a good idea to compare floating numbers (either single or double) with `==`

File: tests/customed_header.h [hunk 6]
```diff
@@ -1,5 +1,5 @@
 
-static inline void func(CField *f, float *lon, float *dt)
+static inline void func(CField *f, double *lon, float *dt)
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply tests/customed_header.h]: We only test the `double` dtype in this test

File: parcels/particleset.py (hunk 1 above)
Comment (delandmeterp) [reply parcels/particleset.py]: Done

File: tests/test_advection.py [hunk 7]
```diff
@@ -307,8 +307,8 @@ def fieldset_decaying(xdim=100, ydim=100, maxtime=delta(hours=6)):
     ('RK45', 1e-5)])
 def test_decaying_eddy(fieldset_decaying, mode, method, rtol, npart=1):
     fieldset = fieldset_decaying
-    lon = np.linspace(12000, 21000, npart, dtype=np.float32)
-    lat = np.linspace(12500, 12500, npart, dtype=np.float32)
+    lon = np.linspace(12000, 21000, npart)
+    lat = np.linspace(12500, 12500, npart)
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply tests/test_advection.py]: No. It was breaking only if we were comparing in python the value of a float with a double

File: parcels/particleset.py [hunk 8]
```diff
@@ -28,10 +29,13 @@ class ParticleSet(object):
     :param depth: Optional list of initial depth values for particles. Default is 0m
     :param time: Optional list of initial time values for particles. Default is fieldset.U.grid.time[0]
     :param repeatdt: Optional interval (in seconds) on which to repeat the release of the ParticleSet
+    :param coordinates_var_precision: Floating precision for lon, lat, depth particle coordinates.
```
[the remark this answers is not in the document]
Comment (delandmeterp) [reply parcels/particleset.py]: User generally don't define a dtype to plon and plat. but the dtype is in pset constructor.
+ our default way of defining it.

**Pull request merged**